src/id.ts

```diff
@@ -22,19 +31,38 @@ import { encodeBase58 } from './core/base58.js';
  *
  * @returns base58 encoded v4 UUID
  */
-export function generate(): string {
+export function generate(): Id {
   const uuid = uuidv4();
   const stripped = uuid.replaceAll(/-/g, '');
-  const id = encodeBase58(stripped);
+  const rawId = encodeBase58(stripped);
 
   // In extremely rare occasions the id generator may result in ids that are
   // 21 characters instead of 22. Theoretically the smallest length the id can
   // generate is 16 characters, but only in specifically engineered cases.
   //
   // If this occurs we can generate again until we get a valid id.
-  if (id.length === 22) {
-    return id;
+  if (rawId.length === 22) {
+    return Id(rawId);
   }
 
   return generate();
 }
+
+export function isValid(id: string): boolean {
+  if (id.length !== 22 && id.length !== 21) {
+    return false;
+  }
+
+  try {
+    const decoded = decodeBase58ToUUID(id);
+    return uuidValidate(decoded);
+  } catch (error) {
+    return false;
+  }
+}
+
+export function assertValid(id: string) {
```

Comment:
Might be useful to pass a specific error message to the Error so consumers get more context as to why the error happened.

Reply:
Gave an example here https://github.com/graphprotocol/grc-20-ts/pull/42/files#r1972426106

Reply:
👍 will do this in a follow up PR